pkg/util/watch/until.go

```diff
@@ -0,0 +1,403 @@
+				a.Namespace, b.Namespace)
+		}
+		c.Namespace = a.Namespace
+	case a.Namespace != "":
+		c.Namespace = a.Namespace
+	case b.Namespace != "":
+		c.Namespace = b.Namespace
+	}
+
+	switch {
+	case a.FieldSelector != nil && b.FieldSelector != nil:
+		c.FieldSelector = fields.AndSelectors(a.FieldSelector, b.FieldSelector)
+	case a.FieldSelector != nil:
+		c.FieldSelector = a.FieldSelector
+	case b.FieldSelector != nil:
+		c.FieldSelector = b.FieldSelector
+	}
+
+	switch {
+	case a.LabelSelector != nil && b.LabelSelector != nil:
+		bReqs, selectable := b.LabelSelector.Requirements()
+		if selectable {
+			c.LabelSelector = a.LabelSelector.Add(bReqs...)
+		} else {
+			c.LabelSelector = a.LabelSelector
+		}
+	case a.FieldSelector != nil:
+		c.LabelSelector = a.LabelSelector
+	case b.FieldSelector != nil:
+		c.LabelSelector = b.LabelSelector
+	}
+
+	switch {
+	case a.Limit > 0 && b.Limit > 0:
+		if a.Limit != b.Limit {
+			return nil, errors.Errorf("cannot merge two different limits: %d & %d",
+				a.Limit, b.Limit)
+		}
+		c.Limit = a.Limit
+	case a.Limit > 0:
+		c.Limit = a.Limit
+	case b.Limit > 0:
+		c.Limit = b.Limit
+	}
+
+	switch {
+	case a.Continue != "" && b.Continue != "":
+		if a.Continue != b.Continue {
+			return nil, errors.Errorf("cannot merge two different continue tokens: %s & %s",
+				a.Continue, b.Continue)
+		}
+		c.Continue = a.Continue
+	case a.Continue != "":
+		c.Continue = a.Continue
+	case b.Continue != "":
+		c.Continue = b.Continue
+	}
+
+	switch {
+	case a.Raw != nil && b.Raw != nil:
+		// TODO: It should be possible to merge raw ListOptions, but we don't need it yet
+		return nil, errors.Errorf("not yet implemented: merging two different raw ListOptions: %+v & %+v",
+			a.Raw, b.Raw)
+	case a.Raw != nil:
+		c.Raw = a.Raw
+	case b.Raw != nil:
+		c.Raw = b.Raw
+	}
+
+	return &c, nil
+}
+
+// ConvertListOptions converts from `metav1.ListOptions` to `client.ListOptions`.
+//
+// For the reverse, see `client.ListOptions.AsListOptions()`.
+func ConvertListOptions(mOpts *metav1.ListOptions) (*client.ListOptions, error) {
+	if mOpts == nil {
+		return nil, nil
+	}
+	cOpts := client.ListOptions{}
+
+	if mOpts.FieldSelector != "" {
+		fs, err := fields.ParseSelector(mOpts.FieldSelector)
+		if err != nil {
+			return &cOpts, errors.Wrap(err, "failed to parse FieldSelector")
+		}
+		cOpts.FieldSelector = fs
+	}
+	if mOpts.LabelSelector != "" {
+		ls, err := labels.Parse(mOpts.LabelSelector)
+		if err != nil {
+			return &cOpts, errors.Wrap(err, "failed to parse LabelSelector")
+		}
+		cOpts.LabelSelector = ls
+	}
+	if mOpts.Limit > 0 {
+		cOpts.Limit = mOpts.Limit
+	}
+	if mOpts.Continue != "" {
+		cOpts.Continue = mOpts.Continue
+	}
+
+	// Pass raw metav1.ListOptions, in case some options are specified that
+	// aren't fields of client.ListOptions.
+	// The fields on client.ListOptions will take precedence over the raw
+	// metav1.ListOptions fields.
+	cOpts.Raw = mOpts
```

Comment:
Seems inconsistent that we support raw `ListOptions` here but not in the above function

Reply:
We support ONE of the ListOptions containing Raw fields, just not both.